### PR TITLE
Support Lucene/Solr 6.2.1

### PR DESCRIPTION
**Pull request by hkazuakey**
Support Lucene/Solr 6.2.1 and add a CharFilter that applies character normalization based on UNICODE standard.

### CODE REVIEW

File: src/java/net/java/sen/dictionary/Morpheme.java [hunk 1]
```diff
@@ -348,7 +348,12 @@ public String toString() {
     partOfSpeechData.append((pronunciations.size() > 0) ? pronunciations.get(0) : "null");
     return partOfSpeechData.toString();
   }
-  
+
+  @Override
+  public int hashCode() {
```
Comment (johtani) [inline src/java/net/java/sen/dictionary/Morpheme.java]: I'm not sure why it needs. Does it need for lucene-6.2.1?

File: src/java/net/java/sen/dictionary/Reading.java [hunk 2]
```diff
@@ -48,7 +50,12 @@ public boolean equals(Object object) {
   public String toString() {
     return "Reading:{" + start + ":" + length + ":" + text + "}";
   }
-  
+
+  @Override
+  public int hashCode() {
```
Comment (johtani) [inline src/java/net/java/sen/dictionary/Reading.java]: same here

File: src/java/net/java/sen/dictionary/Token.java [hunk 3]
```diff
@@ -220,6 +222,11 @@ public boolean equals(Object object) {
   public String toString() {
     return getSurface();
   }
+
+  @Override
+  public int hashCode() {
```
Comment (johtani) [inline src/java/net/java/sen/dictionary/Token.java]: same

File: src/java/org/apache/lucene/analysis/gosen/GosenCharacterNormalizeFilter.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import com.ibm.icu.text.Normalizer2;
+import org.apache.lucene.analysis.charfilter.BaseCharFilter;
+import org.apache.lucene.analysis.util.RollingCharBuffer;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.io.StringReader;
+
+
+public class GosenCharacterNormalizeFilter extends BaseCharFilter {
```
Comment (johtani) [inline src/java/org/apache/lucene/analysis/gosen/GosenCharacterNormalizeFilter.java]: I'm wondering if it should be GosenNomalizerCharFilter. What do you think?

File: src/test/org/apache/lucene/analysis/gosen/TestGosenCharacterNormalizeFilter.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+  // Nothing to change - Just make sure normalizer is working.
+  @Test
+  public void testNothingChange() throws Exception {
+    checkToken("x", "x");
+  }
+
+  // Normalize Full-width Latin to Half-width Latin
+  @Test
+  public void test01() throws Exception {
```
Comment (johtani) [inline src/test/org/apache/lucene/analysis/gosen/TestGosenCharacterNormalizeFilter.java]: I think the clearer name is good.

File: src/test/org/apache/solr/analysis/TestGosenCharacterNormalizeFilterFactory.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  /**
+   * Test nfkc nomarlization that normalize characters to half-width Katakana
+   * from full-width Katakana.
+   *
+   * @throws IOException
+   */
+  @Test
+  public void testNormalization01() throws IOException {
```
Comment (johtani) [inline src/test/org/apache/solr/analysis/TestGosenCharacterNormalizeFilterFactory.java]: It should be the exact name we use

File: src/java/net/java/sen/dictionary/Morpheme.java (hunk 1 above)
Comment (hkazuakey) [inline src/java/net/java/sen/dictionary/Morpheme.java]: It is not required by Lucene 6.2.1. I just wanted to suppress the warning message like the following:
```
[javac] lucene-gosen/src/java/net/java/sen/dictionary/Reading.java:28: warning: [overrides] Class Reading overrides equals, but neither it nor any superclass overrides hashCode method
```
I don't think this method should be included in this change, so we can eliminate this addition.

File: src/java/org/apache/lucene/analysis/gosen/GosenCharacterNormalizeFilter.java (hunk 4 above)
Comment (hkazuakey) [inline src/java/org/apache/lucene/analysis/gosen/GosenCharacterNormalizeFilter.java]: OK. I don't have any strong preference and opinion. To be honest, I don't think I have a good sense of naming :)
So, I will change its class name to GosenNomalizerCharFilter.

File: src/test/org/apache/lucene/analysis/gosen/TestGosenCharacterNormalizeFilter.java (hunk 5 above)
Comment (hkazuakey) [inline src/test/org/apache/lucene/analysis/gosen/TestGosenCharacterNormalizeFilter.java]: OK. I agree. I will update them.

**Comment by hkazuakey** [comment]
@johtani Thank you for your feedback and comments!  I will update based on your feedback.

**Comment by hkazuakey** [comment]
Sorry for this late response. Based on your feedback, I have updated the codes and pushed the latest one.

**Pull request merged**